tournamentcontrol/competition/tests/test_utils.py

```diff
@@ -124,3 +129,161 @@ def test_sum_dict_integer(self):
+        """
+        Test that stage_group_position works correctly when the pool exists.
+        """
+        # Create a division with multiple stages
+        division = DivisionFactory.create()
+        stage1 = StageFactory.create(division=division, order=1)
+        stage2 = StageFactory.create(division=division, order=2, follows=stage1)
+        stage3 = StageFactory.create(division=division, order=3, follows=stage1)
+
+        # Stage1 has 2 pools
+        pool1 = StageGroupFactory.create(stage=stage1)
+        pool2 = StageGroupFactory.create(stage=stage1)
+
+        # Test formulas both with and without explicit stage reference
+        test_cases = [
+            # (from_stage, with_stage_formula, without_stage_formula, expected_stage, expected_group, expected_position)
+            (stage2, "S1G1P1", "G1P1", stage1, pool1, 1),
+            (stage2, "S1G2P3", "G2P3", stage1, pool2, 3),
+            (stage3, "S1G1P1", "G1P1", stage1, pool1, 1),
+            (stage3, "S1G2P3", "G2P3", stage1, pool2, 3),
+        ]
+
+        for (
+            from_stage,
+            with_stage_formula,
+            without_stage_formula,
+            expected_stage,
+            expected_group,
+            expected_position,
+        ) in test_cases:
+            with self.subTest(from_stage=from_stage, formula=with_stage_formula):
+                # Test with explicit stage reference (e.g., "S1G1P1")
+                result_stage, result_group, result_position = (
+                    utils.stage_group_position(from_stage, with_stage_formula)
+                )
+                self.assertEqual(result_stage, expected_stage)
+                self.assertEqual(result_group, expected_group)
+                self.assertEqual(result_position, expected_position)
+
+            with self.subTest(from_stage=from_stage, formula=without_stage_formula):
+                # Test without explicit stage reference (e.g., "G1P1")
+                # This should use from_stage.comes_after
+                result_stage, result_group, result_position = (
+                    utils.stage_group_position(from_stage, without_stage_formula)
+                )
+                self.assertEqual(result_stage, expected_stage)
+                self.assertEqual(result_group, expected_group)
+                self.assertEqual(result_position, expected_position)
+
+    def test_edit_stage_view_with_undecided_teams_referencing_missing_pools(self):
+        """
+        Integration test that exercises the edit_stage admin view where UndecidedTeam
+        objects reference pools that don't exist, reproducing the original IndexError.
+
+        This test ensures that the admin view handles the error gracefully after
+        the fix, rather than crashing with an IndexError when rendering templates
+        that access UndecidedTeam.title or UndecidedTeam.choices properties.
+        """
+        from django.contrib.auth.models import User
+        from django.test import Client
```

Comment:
Please refactor and use the best practices of [`django-test-plus`](https://django-test-plus.readthedocs.io/).